### PR TITLE
[WIP][don't merge] Dev:1st impl for integration test

### PR DESCRIPTION
**Pull request by SamYuan1990**
Details for this PR
e2e folder with end to end testing, will run with two models, base on environment `kepler_address`.
1. build kepler during test process and run it targets on local.
2. run test case targets on cluster after cluster port forwarding.

Signed-off-by: Sam Yuan <yy19902439@126.com>

### CODE REVIEW

**Comment by SamYuan1990** [comment]
@marceloamaral , if I am correct, in integration test, we will deploy kepler with port 9102 right? could you please help with error https://github.com/sustainable-computing-io/kepler/actions/runs/3174554723/jobs/5171489975#step:8:28

File: cmd/exporter.go [hunk 1]
```diff
@@ -77,10 +77,14 @@ func main() {
 	if err != nil {
 		klog.Fatalf("%s", fmt.Sprintf("failed to create collector: %v", err))
 	}
-	err = newCollector.Attach()
-	if err != nil {
+	_ = newCollector.Attach()
+	/*
+		remove this part of error check for now, as it binds with bcc library
```
Comment (SamYuan1990) [inline cmd/exporter.go:82]: @rootfs , could you please suggest/advise?

**Comment by SamYuan1990** [comment]
Sorry, for asking for help on this WIP pr.

File: cmd/exporter.go [hunk 2]
```diff
@@ -77,10 +77,14 @@ func main() {
...
+		if instance wihout bcc, start will fail.
+		if err != nil {
 		klog.Fatalf("%s", fmt.Sprintf("failed to attach : %v", err))
```
Comment (rootfs) [inline cmd/exporter.go:85]: why this failed?

File: cmd/exporter.go (hunk 1 above)
Comment (SamYuan1990) [reply cmd/exporter.go:82]: https://github.com/sustainable-computing-io/kepler/actions/workflows/unit_test_mac.yml
https://github.com/sustainable-computing-io/kepler/actions/workflows/unit_test.yml
for those two jobs we don't have bcc.

the reason in deep for the failure is about the reader logic.
https://github.com/sustainable-computing-io/kepler/blob/main/pkg/collector/reader.go#L106-L109
and some null pointer return by `pkg/attacher/bcc_attacher_stub.go` if bcc not installed.

I suppose there are two ways, either we make our code robust(for rest api handler part) or we make our code decouple with bcc(if we can't decouple with bcc, we force ourself bind with it.)

Comment (rootfs) [reply cmd/exporter.go:82]: @SamYuan1990 if this is doable, maybe have yet another bcc attacher stub implementation for integration test and in the stub function, return certain mock data
https://github.com/sustainable-computing-io/kepler/blob/main/pkg/attacher/bcc_attacher_stub.go#L42-L52
e.g. 
```go
func (table *Table) Iter() *TableIterator {
	return &TableIterator{mock_data}
}

func (it *TableIterator) Next() bool {
	return false
}

func (it *TableIterator) Leaf() []byte {
	return mock_data
}
```

File: pkg/collector/reader.go [hunk 3]
```diff
@@ -104,8 +104,12 @@ func (c *Collector) resetCurrValue() {
 
 // resetBPFTables reset BPF module's tables
 func (c *Collector) resetBPFTables() {
-	c.modules.Table.DeleteAll()
-	c.modules.TimeTable.DeleteAll()
+	if c.modules != nil && c.modules.Table != nil {
```
Comment (SamYuan1990) [inline pkg/collector/reader.go:107]: @rootfs , do you know where we init modules and tables?

File: pkg/power/acpi/util_linux.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
+//go:build linux
+// +build linux
```
Comment (SamYuan1990) [inline pkg/power/acpi/util_linux.go:2]: @rootfs ,those path are linux specific, so we have to split at here for different device?

File: pkg/attacher/bcc_attacher_stub.go [hunk 5]
```diff
@@ -66,7 +62,7 @@ var (
 )
 
 func AttachBPFAssets() (*BpfModuleTables, error) {
-	return nil, fmt.Errorf("no bcc build tag")
+	return nil, nil
```
Comment (SamYuan1990) [inline pkg/attacher/bcc_attacher_stub.go:65]: @rootfs , here is the pre defined error for without bcc tag.

**Comment by SamYuan1990** [comment]
@rootfs , @marceloamaral , it seems we got error at https://github.com/sustainable-computing-io/kepler/actions/runs/3180212481/jobs/5183521776#step:7:17 could you please advise?

**Comment by SamYuan1990** [comment]
topic may related https://github.com/kubernetes-sigs/kind/search?q=failed+to+execute+portforward+in+network+namespace+%22host%22&type=issues

**Comment by SamYuan1990** [comment]
> > topic may related https://github.com/kubernetes-sigs/kind/search?q=failed+to+execute+portforward+in+network+namespace+%22host%22&type=issues
> 
> Could [`ClusterIP: None`](https://github.com/sustainable-computing-io/kepler/blob/main/manifests/kubernetes/deployment.yaml#L126) be causing trouble?

I give a try by a new commit, but seems not the root cause.
from the log, I suppose there some error happened during ip v4 ip v6 transform.

File: pkg/power/acpi/util_linux.go (hunk 4 above)
Comment (rootfs) [reply pkg/power/acpi/util_linux.go:2]: @marceloamaral any idea? maybe a stub here too?

File: pkg/collector/reader.go (hunk 3 above)
Comment (rootfs) [reply pkg/collector/reader.go:107]: the tables are auto filled; the delete functions remove existing table entries, the ebpf program will fill the tables again.

**Comment by BenTheElder** [comment]
> @rootfs , I am now try to force KIND use IPv4 to avoid:
> https://kind.sigs.k8s.io/docs/user/known-issues/#ipv6-port-forwarding

That issue is only relevant when you are asking kind to forward ports from the host to the node, which you don't appear to be using here.

When you are using the Kubernetes port forward API to tunnel traffic through the API server to a workload, that is completely unrelated.


AFAICT, the kubernetes-api-level port forward is failing to connect to localhost:9102 in the pod namespace with neither ipv4 nor ipv6 connections succeeding. It's not clear why, other than the connection is refused.

> E1004 08:02:47.046060 7990 portforward.go:406] an error occurred forwarding 9102 -> 9102: error forwarding port 9102 to pod 2b02221f9f4b01c499fabccd3d35d62e5591d6b1303d9[18](https://github.com/sustainable-computing-io/kepler/actions/runs/3180383090/jobs/5183888088#step:7:19)294587bd80d23e38f, uid : failed to execute portforward in network namespace "host": failed to connect to localhost:9102 inside namespace "2b022[21](https://github.com/sustainable-computing-io/kepler/actions/runs/3180383090/jobs/5183888088#step:7:22)f9f4b01c499fabccd3d35d62e5591d6b1303d918294587bd80d[23](https://github.com/sustainable-computing-io/kepler/actions/runs/3180383090/jobs/5183888088#step:7:24)e38f", IPv4: dial tcp4 1[27](https://github.com/sustainable-computing-io/kepler/actions/runs/3180383090/jobs/5183888088#step:7:28).0.0.1:9102: connect: connection refused IPv6 dial tcp6 [::1]:9102: connect: connection refused

Are we sure there's an application listening at 127.0.0.1:9102 when we start the port forward?

More discussion in https://github.com/kubernetes-sigs/kind/issues/2953

**Comment by rootfs** [comment]
@BenTheElder Thank you Ben for the help! The application should be running and listening to the port (in [this](https://github.com/sustainable-computing-io/kepler/blob/main/manifests/kubernetes/deployment.yaml#L119) Service).

**Comment by BenTheElder** [comment]
port-forward forwards to pods (not services), it will attempt to open a connection from within the pod's namespace, the tunnel looks roughly like:

kubectl <=> API server <=> kubelet <=> CRI <=> socket opened within target pod. IIRC kubernetes will always do this on `localhost`. The specific implementation in this case will be containerd's.

My KIND co-maintainer aojea previously patched this, because it used to literally use `socat` and supported IPv4 only. The current implementation has "happy eyeballs" IPv4+IPv6 (because Go) IIRC.

I'm noticing that the pod seems to be configured in hostNetwork and looks like the process should be bound to 0.0.0.0 ...

I'm not sure if we / Kubernetes have tests for hostNetwork port forward, it's possible that's broken in KIND (or containerd), otherwise unless the app isn't listening yet or the wrong pod is selected, I'm at a loss ...

**Comment by rootfs** [comment]
Good info @BenTheElder! So based on your feedback, a more appropriate CI deployment is to use pod network. @SamYuan1990 can you give that a try? Thanks!

**Comment by BenTheElder** [comment]
Instead of kubectl port forward, you could try configuring a port forward in kind and use a matching nodePort, similar to https://kind.sigs.k8s.io/docs/user/ingress/

`kubectl port forward` is more of a debugging tool and is not the most robust, though i'd still expect it to work here 🤔 , but as a general rule I wouldn't use it for anything load bearing with any cluster, even if you use the API directly.

Something like:
```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- extraPortMappings:
  - containerPort: 30001
    hostPort: 30001
    protocol: TCP
```


Or actually since the pod appears(?) to be in hostNetwork and you have a single node:
```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- extraPortMappings:
  - containerPort: 9102
    hostPort: 9102
    protocol: TCP
```

> if I am right, github agent doesn't support ipv6 for now.
> https://github.com/actions/runner-images/issues/668

We run some of our CI in GitHub actions and lack of IPv6 is not an issue, the ipv6 port forward connect is a red-herring, it's portforward is trying both ipv4 and ipv6 and both are failing. You can use kind / kubernetes without ipv6, if anything optional ipv6 in KIND involved a _lot_ of fixing tests and Kubernetes, and then again in conjunction with the folks working on  dualstack.

**Comment by SamYuan1990** [comment]
> Instead of kubectl port forward, you could try configuring a port forward in kind and use a matching nodePort, similar to https://kind.sigs.k8s.io/docs/user/ingress/
> 
> `kubectl port forward` is more of a debugging tool and is not the most robust, though i'd still expect it to work here 🤔 , but as a general rule I wouldn't use it for anything load bearing with any cluster, even if you use the API directly.
> 
> Something like:
> 
> ```yaml
> kind: Cluster
> apiVersion: kind.x-k8s.io/v1alpha4
> nodes:
> - extraPortMappings:
>   - containerPort: 30001
>     hostPort: 30001
>     protocol: TCP
> ```
> 
> Or actually since the pod appears(?) to be in hostNetwork and you have a single node:
> 
> ```yaml
> kind: Cluster
> apiVersion: kind.x-k8s.io/v1alpha4
> nodes:
> - extraPortMappings:
>   - containerPort: 9102
>     hostPort: 9102
> ```
> 
> I.E. adds `-p 127.0.0.1:9102:9102/TCP` to running the kind node.
> 
> > if I am right, github agent doesn't support ipv6 for now.
> > [actions/runner-images#668](https://github.com/actions/runner-images/issues/668)
> 
> We run some of our CI in GitHub actions and lack of IPv6 is not an issue, the ipv6 port forward connect is a red-herring, it's portforward is trying both ipv4 and ipv6 and both are failing. You can use kind / kubernetes without ipv6, if anything optional ipv6 in KIND involved a _lot_ of fixing tests and Kubernetes, and then again in conjunction with the folks working on dualstack.

let me try

**Comment by SamYuan1990** [comment]
@rootfs , @marceloamaral , take a look at https://github.com/sustainable-computing-io/kepler/actions/runs/3201921603/jobs/5230372207#step:8:13 pls

**Comment by SamYuan1990** [comment]
some updates:
- tried with https://kind.sigs.k8s.io/docs/user/configuration/#extra-port-mappings but, `The Service "kepler-exporter" is invalid: spec.ports[0].nodePort: Forbidden: may not be used when `type` is 'ClusterIP'` should we use `ClusterIP` together with extra-port-mappings?
- fixed one panic error, but why pod in panic passed `kubectl rollout status daemonset kepler-exporter -n kepler --timeout 180s` ?  question here as the panic will make the service down, which breaks livenessProbe. 
- once panic been fixed, the port forward works. @_@

**Comment by BenTheElder** [comment]
> The Service "kepler-exporter" is invalid: spec.ports[0].nodePort: Forbidden: may not be used when type is 'ClusterIP' should we use ClusterIP together with extra-port-mappings?

nodePort can only be set on a service of type [NodePort](https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport). shouldn't be related to the node level port mappings.

I think you can skip the service though and just map 9102 from the host to the kind node since the pod appears to be in hostNetwork.

**Comment by marceloamaral** [comment]
I've been using port-forward with no problem on a kind cluster.
So probably the problem here was just related to the panic error.

The command `kubectl rollout status daemonset kepler-exporter -n kepler --timeout 180s` verifies if the container has been started, but will not check if the container failed shortly after starting....

Maybe we should sleep and then check for failed containers....

**Comment by rootfs** [comment]
@SamYuan1990 now just a couple of lint error, ready to merge

**Comment by SamYuan1990** [comment]
> @SamYuan1990 now just a couple of lint error, ready to merge

 no no no.... I am going to bed... and don't merge it for now. please take a look at #283 1st.

**Comment by SamYuan1990** [comment]
close and move to #292 without issue. thanks guys.

**Pull request closed without merging**